ipn/ipnlocal/network-lock.go

```diff
@@ -122,6 +136,67 @@ func (b *LocalBackend) tkaFilterNetmapLocked(nm *netmap.NetworkMap) {
 	}
 }
 
+// rotationTracker determines the set of node keys that are made obsolete by key
+// rotation.
+//   - for each SigRotation signature, all previous node keys referenced by the
+//     nested signatures are marked as obsolete.
+//   - if there are multiple SigRotation signatures tracing back to the same
+//     wrapping pubkey (e.g. if a node is cloned with all its keys), we keep
+//     just one of them, marking the others as obsolete.
+type rotationTracker struct {
+	// obsolete is the set of node keys that are obsolete due to key rotation.
+	obsolete set.Set[key.NodePublic]
+
+	// byWrappingKey keeps track of rotation details per wrapping pubkey.
+	byWrappingKey map[string][]sigRotationDetails
+}
+
+// sigRotationDetails holds information about a node key signed by a SigRotation.
+type sigRotationDetails struct {
+	np          key.NodePublic
+	numPrevKeys int
+}
+
+// addRotationDetails records the rotation signature details for a node key.
+func (r *rotationTracker) addRotationDetails(np key.NodePublic, d *tka.RotationDetails) {
+	r.obsolete.Make()
+	r.obsolete.AddSlice(d.PrevNodeKeys)
+	rd := sigRotationDetails{
+		np:          np,
+		numPrevKeys: len(d.PrevNodeKeys),
+	}
+	if r.byWrappingKey == nil {
+		r.byWrappingKey = make(map[string][]sigRotationDetails)
+	}
+	wp := string(d.WrappingPubkey)
+	r.byWrappingKey[wp] = append(r.byWrappingKey[wp], rd)
+}
+
+// obsoleteKeys returns the set of node keys that are obsolete due to key rotation.
+func (r *rotationTracker) obsoleteKeys() set.Set[key.NodePublic] {
+	for _, v := range r.byWrappingKey {
+		// Sort by decreasing number of previous keys.
+		sort.SliceStable(v, func(i, j int) bool {
+			return v[i].numPrevKeys > v[j].numPrevKeys
+		})
+		skipped := false
+		for _, rd := range v {
+			if r.obsolete.Contains(rd.np) {
+				continue
+			}
+			// Reached the node that has the largest number of previous keys in
+			// its signature. Skip it, and add the rest to the list of obsolete
+			// keys.
+			if !skipped {
+				skipped = true
+				continue
+			}
+			r.obsolete.Add(rd.np)
+		}
```

Comment:
this took me a while to understand.
can this be something like this instead?
```go
// The first key in v is one with the longest chain of previous keys, so it must be the newest one.
// Mark all older keys (with shorter chains) as obsolete.
for _, rd := range v[1:] {
	r.obsolete.Add(rd.np)
}
```

Reply:
I am also excluding keys that have already been marked as obsolete (because one of the other signatures rotated them out) even if they have a longer chain length. This should not happen in practice at the moment, I believe, but if at some point we start compacting long chains of rotated signatures, "longest" will not necessarily mean "latest".

To rephrase, I am treating "rotated out" as a stronger signal, and the chain length as a secondary signal.

I've moved this logic to the sorting function to make it a bit more explicit, and added more details to the comments.